mailpoet/lib/WPCOM/DotcomLicenseProvisioner.php

```diff
@@ -0,0 +1,134 @@
+ */
+class DotcomLicenseProvisioner {
+  const EVENT_TYPE_PROVISION_LICENSE = 'provision_license';
+
+  /** @var LoggerFactory */
+  private $loggerFactory;
+
+  /** @var Settings */
+  private $settings;
+
+  /** @var Services */
+  private $services;
+
+  public function __construct(
+    LoggerFactory $loggerFactory,
+    Settings $settings,
+    Services $services
+  ) {
+    $this->loggerFactory = $loggerFactory;
+    $this->settings = $settings;
+    $this->services = $services;
+  }
+
+  /**
+   * Returns true if in the context of WordPress.com Atomic platform.
+   *
+   * @return bool
+   */
+  public function isAtomicPlatform(): bool {
+    return defined('IS_ATOMIC') && IS_ATOMIC && defined('ATOMIC_CLIENT_ID') && (ATOMIC_CLIENT_ID === '2');
+  }
+
+  /**
+   * Activates MSS and adds API key for subscriptions purchased from WP.com Marketplace.
+   *
+   * @param bool $result
+   * @param array $licensePayload
+   * @param string $eventType
+   * @return bool|WP_Error
+   */
+  public function provisionLicense(bool $result, array $licensePayload, string $eventType) {
+    if (!$this->isAtomicPlatform() || $eventType !== self::EVENT_TYPE_PROVISION_LICENSE) {
+      return $result;
+    }
+
+    $apiKey = $this->getKeyFromPayload($licensePayload);
+    if (is_wp_error($apiKey)) {
+      $this->loggerFactory->getLogger(LoggerFactory::TOPIC_PROVISIONING)->error(
+        'key was not found in license payload');
+      return $apiKey;
+    }
+
+    return $this->activateMSS($apiKey);
+  }
+
+  /**
+   * Returns API key from license payload.
+   * @param array $licensePayload
+   * @return string|WP_Error
+   */
+  private function getKeyFromPayload(array $licensePayload) {
+    if (isset($licensePayload['apiKey']) && is_string($licensePayload['apiKey'])) {
+      return $licensePayload['apiKey'];
+    }
+
+    return new WP_Error('invalid_license_payload', 'Invalid license payload: Missing API key.');
+  }
+
+  /**
+   * Saves the API key and activates MSS.
+   * @param string $apiKey
+   * @return true|WP_Error
+   */
+  public function activateMSS(string $apiKey) {
+    $response = $this->settings->setupMSS($apiKey);
+    if ($response instanceof ErrorResponse) {
+      $this->loggerFactory->getLogger(LoggerFactory::TOPIC_PROVISIONING)->error(
+        'Setting sending method and key failed',
+        ['$response' => $response]
+      );
+      return new WP_Error('Provisioning failed setting the data', $this->concatMessages($response));
+    }
+
+    // This is necessary if the key changed but the sending method was already set to MailPoet
+    $response = $this->services->refreshMSSKeyStatus();
+    if ($response instanceof ErrorResponse) {
+      $this->loggerFactory->getLogger(LoggerFactory::TOPIC_PROVISIONING)->error(
+        'Refreshing the key failed',
+        ['$response' => $response]
+      );
+      return new WP_Error('Provisioning failed activating the data', $this->concatMessages($response));
```

Comment:
The error message is the same for MSS and Premium key. I think it would be nice to distinguish these two. Maybe `Provisioning failed to verify api key access for MSS/premium` ?